clients/src/main/java/org/apache/kafka/clients/MetadataCache.java

```diff
@@ -0,0 +1,192 @@
+ */
+class MetadataCache {
+    private final String clusterId;
+    private final List<Node> nodes;
+    private final Set<String> unauthorizedTopics;
+    private final Set<String> invalidTopics;
+    private final Set<String> internalTopics;
+    private final Node controller;
+    private final Map<TopicPartition, PartitionInfoAndEpoch> partitionsByTopicPartition;
+
+    private Cluster clusterInstance;
+
+    MetadataCache(String clusterId,
+                  List<Node> nodes,
+                  Collection<PartitionInfoAndEpoch> partitions,
+                  Set<String> unauthorizedTopics,
+                  Set<String> invalidTopics,
+                  Set<String> internalTopics,
+                  Node controller) {
+        this(clusterId, nodes, partitions, unauthorizedTopics, invalidTopics, internalTopics, controller, null);
+    }
+
+    MetadataCache(String clusterId,
+                  List<Node> nodes,
+                  Collection<PartitionInfoAndEpoch> partitions,
+                  Set<String> unauthorizedTopics,
+                  Set<String> invalidTopics,
+                  Set<String> internalTopics,
+                  Node controller,
+                  Cluster clusterInstance) {
+        this.clusterId = clusterId;
+        this.nodes = nodes;
+        this.unauthorizedTopics = unauthorizedTopics;
+        this.invalidTopics = invalidTopics;
+        this.internalTopics = internalTopics;
+        this.controller = controller;
+
+        this.partitionsByTopicPartition = new HashMap<>(partitions.size());
+        for (PartitionInfoAndEpoch p : partitions) {
+            this.partitionsByTopicPartition.put(new TopicPartition(p.partitionInfo().topic(), p.partitionInfo().partition()), p);
+        }
+
+        if (clusterInstance == null) {
+            computeClusterView();
+        } else {
+            this.clusterInstance = clusterInstance;
+        }
+    }
+
+    /**
+     * Return the cached PartitionInfo iff it was for the given epoch
+     * @param topicPartition
+     * @param epoch
+     * @return
+     */
+    Optional<PartitionInfo> getPartitionInfoHavingEpoch(TopicPartition topicPartition, int epoch) {
+        PartitionInfoAndEpoch infoAndEpoch = partitionsByTopicPartition.get(topicPartition);
+        if (infoAndEpoch == null) {
+            return Optional.empty();
+        } else {
+            if (infoAndEpoch.epoch() == epoch) {
+                return Optional.of(infoAndEpoch.partitionInfo());
+            } else {
+                return Optional.empty();
+            }
+        }
+    }
+
+    Optional<PartitionInfo> getPartitionInfo(TopicPartition topicPartition) {
+        return Optional.ofNullable(partitionsByTopicPartition.get(topicPartition))
+                .map(PartitionInfoAndEpoch::partitionInfo);
+    }
+
+    void removePartitionInfo(TopicPartition topicPartition) {
+        partitionsByTopicPartition.entrySet().removeIf(entry -> entry.getKey().equals(topicPartition));
```

Comment:
Isn't this just `partitionsByTopicPartition.remove(topicPartition)`? By the way, the name is a bit confusing. How about `metadataByPartition`?